compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala

```diff
@@ -0,0 +1,914 @@
+         * by explicitouter, while the other capture params are added by
+         * lambdalift (between elimErasedValueType and now).
+         *
+         * scalac: Note that lambdalift creates new symbols even for parameters that
+         * are not the result of lambda lifting, but it preserves their
+         * `name`s.
+         */
+
+        val hasOuterParam = {
+          paramInfosAtElimEVT.size == paramInfosAtElimRepeated.size + 1 &&
+          paramNamessAtElimEVT.flatten.head == nme.OUTER
+        }
+        assert(
+            hasOuterParam || paramInfosAtElimEVT.size == paramInfosAtElimRepeated.size,
+            s"Found ${paramInfosAtElimRepeated.size} params entering elimRepeated but " +
+            s"${paramInfosAtElimEVT.size} params entering elimErasedValueType for " +
+            s"lifted constructor symbol ${sym.fullName}")
+
+        val startOfFormalParams = paramNamessNow.flatten.indexOfSlice(paramNamessAtElimRepeated.flatten)
+        val formalParamCount = paramInfosAtElimRepeated.size
+
+        val nonOuterParamInfosAtElimEVT =
+          if (hasOuterParam) paramInfosAtElimEVT.tail
+          else paramInfosAtElimEVT
+        val formalParams = buildFormalParams(paramSymsAtElimRepeated, paramInfosAtElimRepeated, nonOuterParamInfosAtElimEVT)
+
+        val paramNamesAndInfosNow = paramNamessNow.flatten.zip(paramInfosNow)
+        val (captureParamsFrontNow, restOfParamsNow) = paramNamesAndInfosNow.splitAt(startOfFormalParams)
+        val captureParamsBackNow = restOfParamsNow.drop(formalParamCount)
+
+        def makeCaptureParamDef(nameAndInfo: (TermName, Type)): js.ParamDef = {
+          implicit val pos: Position = sym.span
+          js.ParamDef(freshLocalIdent(nameAndInfo._1.mangledString), NoOriginalName, toIRType(nameAndInfo._2),
+              mutable = false, rest = false)
+        }
+
+        val captureParamsFront = captureParamsFrontNow.map(makeCaptureParamDef(_))
+        val captureParamsBack = captureParamsBackNow.map(makeCaptureParamDef(_))
+
+        /*if (isAnonJSClassConstructor) {
+          // For an anonymous JS class constructor, we put the capture parameters back as formal parameters.
+          val allFormalParams = captureParamsFront.toIndexedSeq ++ formalParams ++ captureParamsBack.toIndexedSeq
+          (allFormalParams, Nil, Nil)
+        } else {*/          (formalParams, captureParamsFront, captureParamsBack)
+        //}
+      }
+    }
+
+    val hasRepeatedParam = params.nonEmpty && params.last.isRepeated
+
+    def pos: SourcePosition = sym.sourcePos
+
+    def exportArgTypeAt(paramIndex: Int): Type = {
+      if (paramIndex < params.length) {
+        params(paramIndex).info
+      } else {
+        assert(hasRepeatedParam, i"$sym does not have varargs nor enough params for $paramIndex")
+        params.last.info
+      }
+    }
+
+    def typeInfo: String = sym.info.toString
+  }
+
+  private sealed abstract class RTTypeTest
```

Comment:
Reviewers: From here until the end of this file, the contents are basically copied verbatim from Scala 2.